Add the Flex EG

jpcima opened the pull request:
Based on PR #386

This implements the Flex EG.

Most important to do:
- implement ARIA `egN_ampeg`
  When this is not set, and `egN` is connected to region amplitude, the region is subject to `ampeg` as its master envelope, and `egN` does not have a release under the default ampeg.
  When `egN_ampeg` ≠ 0, `ampeg` is overriden with `egN`.

More minor things to do:
- implement the modulations of levels and times.
- implement the `egN_dynamic` setting.
  - if `level` is modulated, reconfigure the current stage by setting FlexEG's target level to `level`, source level FlexEG's `currentLevel`, and `currentTime` to 0, `stageEndTime` to `stageEndTime-currentTime`
    (meaning to start over, as if the remainder of the stage is treated as a new curve. it's how ARIA does)
  - if `time` is modulated, update `stageEndTime`

jpcima committed:
Add the Flex EG

jpcima committed:
Some changes of little impact

jpcima committed:
Ensure to update all variables on early return

jpcima committed:
Comment on unimplemented loop opcodes

paulfd commented on src/sfizz/FlexEnvelope.cpp:
```diff
@@ -0,0 +1,206 @@
+
+    // Envelope finished?
+    if (currentStageNumber_ >= desc.points.size()) {
+        fill(out, 0.0f);
```
All envelopes end at 0? Or should it be `currentLevel_`?

jpcima replied:
Yes envelopes end at zero. Try in ARIA:
```
<region> sample=*sine
//eg15_amplitude=100
eg15_ampeg=1
eg15_time1=.1  eg15_level1=.25
eg15_time2=.15  eg15_level2=1
eg15_time3=.2  eg15_level3=.5 eg15_sustain=3
eg15_time4=.4 eg15_level4=1
```

The envelope will return to 0 after the final ramp.
This will create a very noticeable discontinuity.

![Capture du 2020-09-13 07-59-55](https://user-images.githubusercontent.com/17614485/93011519-2225f080-f597-11ea-9b96-9b8e46fd9531.png)

paulfd replied:
Hmm ok, that actually feels a bit strange

paulfd commented on src/sfizz/FlexEnvelope.cpp:
```diff
@@ -0,0 +1,206 @@
+        while ((!stageSustained_ && currentTime_ >= stageTime_) ||
+               (stageSustained_ && isReleased)) {
+            if (!advanceToNextStage()) {
+                fill(out, 0.0f);
```
Same

paulfd committed:
Add a circular test case for the curve

paulfd committed:
Offset the time so that the envelopes "end" on the correct value
The envelope start is always considered to be the value before the current ramp started

paulfd committed:
Only zero the remaining samples when the envelope is finished

paulfd committed:
Add flex eg tests, move test helpers around, and add a more verbose mod matrix graph output

paulfd commented:
I added tests and corrected a couple bugs they uncovered. Other than this it looks ok to me!

jpcima committed:
Print the modulation key more compactly
